### PR TITLE
Feature/implement spring security

### PR DESCRIPTION
**Pull request by Cayman4010**

### CODE REVIEW

File: src/main/java/bookstore/controller/AuthController.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+            description = "Add new user by email and password")
+    public UserResponseDto register(@RequestBody @Valid UserRegistrationRequestDto request)
+            throws RegistrationException {
+        return userService.register(request);
+    }
+
+    @PostMapping("/login")
+    @Operation(summary = "Authentication",
+            description = "Authenticate user by email and password")
+    @ResponseStatus(HttpStatus.OK)
```
Comment (NazarSmith228) [inline src/main/java/bookstore/controller/AuthController.java]: just FYI

`OK` is the default response status, so you can omit it

remember about this in future :)

File: src/main/java/bookstore/controller/AuthController.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    public UserLoginResponseDto login(@RequestBody UserLoginRequestDto requestDto) {
```
Comment (NazarSmith228) [inline src/main/java/bookstore/controller/AuthController.java]: shouldn't you validate request body?

File: src/main/java/bookstore/controller/BookController.java [hunk 3]
```diff
@@ -39,27 +41,31 @@ public List<BookDto> getAll(Pageable pageable) {
...
     @ResponseStatus(HttpStatus.CREATED)
+    @PreAuthorize("hasRole('ADMIN')")
     @PostMapping
     @Operation(summary = "Create a new book", description = "Create a new book")
     public BookDto createBook(@RequestBody @Valid CreateBookRequestDto requestDto) {
         return bookService.save(requestDto);
     }
 
     @ResponseStatus(HttpStatus.OK)
+    @PreAuthorize("hasAnyRole('USER', 'ADMIN')")
```
Comment (NazarSmith228) [inline src/main/java/bookstore/controller/BookController.java]: shouldn't admin user also have a role `USER`?

or you design your app in a way that admin does not have same priviliges as user has?

File: src/main/java/bookstore/dto/user/UserRegistrationRequestDto.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+        @Size(min = 4, max = 60)
+        String password,
+        @NotBlank
+        @Size(min = 4, max = 60)
+        String repeatPassword,
+        @NotBlank
+        String firstName,
+        @NotBlank
+        String lastName,
+        String shippingAddress
```
Comment (NazarSmith228) [inline src/main/java/bookstore/dto/user/UserRegistrationRequestDto.java:23]: shouldn't you validate this property?

File: src/main/java/bookstore/exception/CustomGlobalExceptionHandler.java [hunk 5]
```diff
@@ -41,6 +41,15 @@ protected ResponseEntity<Object> handleEntityNotFoundException(
                 .body(ex.getMessage());
     }
 
+    @ExceptionHandler(RegistrationException.class)
+    protected ResponseEntity<Object> handleRegistrationException(
+            RegistrationException ex
+    ) {
```
Comment (NazarSmith228) [inline src/main/java/bookstore/exception/CustomGlobalExceptionHandler.java]:
```suggestion
    protected ResponseEntity<Object> handleRegistrationException(RegistrationException ex) {
```

File: src/main/java/bookstore/exception/CustomGlobalExceptionHandler.java [hunk 6]
```diff
@@ -41,6 +41,15 @@ protected ResponseEntity<Object> handleEntityNotFoundException(
...
+        return ResponseEntity
+                .status(HttpStatus.BAD_REQUEST)
```
Comment (NazarSmith228) [inline src/main/java/bookstore/exception/CustomGlobalExceptionHandler.java]: you throw this exception when there is duplicate email

in this case you should use HTTP 409 CONFLICT

File: src/main/java/bookstore/model/Book.java [hunk 7]
```diff
@@ -31,5 +33,5 @@ public class Book {
     private String description;
     private String coverImage;
     @Column(nullable = false)
-    private boolean isDeleted = false;
+    private Boolean isDeleted = false;
```
Comment (NazarSmith228) [inline src/main/java/bookstore/model/Book.java]: this makes no sense

you can stick to primitive

File: src/main/java/bookstore/model/User.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.userdetails.UserDetails;
+
+@Entity
+@RequiredArgsConstructor
```
Comment (NazarSmith228) [inline src/main/java/bookstore/model/User.java]: `@NoArgsConstructor`?

File: src/main/java/bookstore/model/User.java [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+    @Column(nullable = false)
+    private String password;
+    @Column(nullable = false)
+    private String firstName;
+    @Column(nullable = false)
+    private String lastName;
+    private String shippingAddress;
+    @Column(nullable = false)
+    private Boolean isDeleted = false;
+    @ManyToMany(fetch = FetchType.EAGER)
```
Comment (NazarSmith228) [inline src/main/java/bookstore/model/User.java]: it is better to add empty lines between field declaration for the sake of readability

File: src/main/java/bookstore/repository/UserRepository.java [hunk 10]
```diff
@@ -0,0 +1,9 @@
+package bookstore.repository;
+
+import bookstore.model.User;
+import java.util.Optional;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface UserRepository extends JpaRepository<User,Long> {
```
Comment (NazarSmith228) [inline src/main/java/bookstore/repository/UserRepository.java:10]: `@Repository` is missing

File: src/main/java/bookstore/model/User.java (hunk 9 above)
Comment (NazarSmith228) [inline src/main/java/bookstore/model/User.java]: eager loading is antipattern - pls don't use it

load roles when necessary via join fetch or entity graph

File: src/main/java/bookstore/security/AuthenticationService.java [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+import org.springframework.stereotype.Service;
+
+@RequiredArgsConstructor
+@Service
+public class AuthenticationService {
+    private final JwtUtil jwtUtil;
+    private final AuthenticationManager authenticationManager;
+
+    public UserLoginResponseDto authenticate(UserLoginRequestDto requestDto) {
+        final Authentication authentication = authenticationManager.authenticate(
```
Comment (NazarSmith228) [inline src/main/java/bookstore/security/AuthenticationService.java]:
```suggestion
        Authentication authentication = authenticationManager.authenticate(
```

File: src/main/java/bookstore/security/CustomUserDetailsService.java [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.stereotype.Service;
+
+@RequiredArgsConstructor
+@Service
+public class CustomUserDetailsService implements UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
```
Comment (NazarSmith228) [inline src/main/java/bookstore/security/CustomUserDetailsService.java]:
```suggestion
    public UserDetails loadUserByUsername(String username) {
```

File: src/main/java/bookstore/security/JwtAuthenticationFilter.java [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+
+    private final JwtUtil jwtUtil;
+    private final UserDetailsService userDetailsService;
+
+    @Override
+    protected void doFilterInternal(
+            HttpServletRequest request,
+            HttpServletResponse response,
+            FilterChain filterChain
+    ) throws ServletException, IOException {
```
Comment (NazarSmith228) [inline src/main/java/bookstore/security/JwtAuthenticationFilter.java]:
```suggestion
            FilterChain filterChain) throws ServletException, IOException {
```

File: src/main/java/bookstore/security/JwtAuthenticationFilter.java [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+            Authentication authentication = new UsernamePasswordAuthenticationToken(
+                    userDetails, null, userDetails.getAuthorities()
+            );
+            SecurityContextHolder.getContext().setAuthentication(authentication);
+        }
+        filterChain.doFilter(request, response);
+    }
+
+    private String getToken(HttpServletRequest request) {
+        String bearerToken = request.getHeader("Authorization");
```
Comment (NazarSmith228) [inline src/main/java/bookstore/security/JwtAuthenticationFilter.java]: use already existing constant - `HttpHeaders.AUTHORIZATION`

File: src/main/java/bookstore/security/JwtAuthenticationFilter.java [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+        if (StringUtils.hasText(bearerToken) && bearerToken.startsWith("Bearer ")) {
+            return bearerToken.substring(BEGIN_TOKEN_INDEX);
```
Comment (NazarSmith228) [inline src/main/java/bookstore/security/JwtAuthenticationFilter.java]: consider creating a constant and:
```suggestion
        if (StringUtils.hasText(bearerToken) && bearerToken.startsWith(BEARER_TYPE)) {
            return bearerToken.substring(BEARER_TYPE.length());
```

File: src/main/java/bookstore/service/user/UserService.java [hunk 16]
```diff
@@ -0,0 +1,9 @@
+package bookstore.service.user;
+
+import bookstore.dto.user.UserRegistrationRequestDto;
+import bookstore.dto.user.UserResponseDto;
+import bookstore.exception.RegistrationException;
+
+public interface UserService {
+    UserResponseDto register(UserRegistrationRequestDto request) throws RegistrationException;
```
Comment (NazarSmith228) [inline src/main/java/bookstore/service/user/UserService.java:8]:
```suggestion
    UserResponseDto register(UserRegistrationRequestDto request);
```

File: src/main/java/bookstore/service/user/impl/UserServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+@Component
+@RequiredArgsConstructor
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final PasswordEncoder passwordEncoder;
+    private final UserMapper userMapper;
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto request)
+            throws RegistrationException {
```
Comment (NazarSmith228) [inline src/main/java/bookstore/service/user/impl/UserServiceImpl.java:27]:
```suggestion
    public UserResponseDto register(UserRegistrationRequestDto request) {
```

File: src/main/java/bookstore/service/user/impl/UserServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+        if (userRepository.findByEmail(request.email()).isPresent()) {
+            throw new RegistrationException("Unable to complete registration");
+        }
+        User user = new User();
+        user.setEmail(request.email());
+        user.setPassword(passwordEncoder.encode(request.password()));
+        user.setFirstName(request.firstName());
+        user.setLastName(request.lastName());
+        user.setShippingAddress(request.shippingAddress());
```
Comment (NazarSmith228) [inline src/main/java/bookstore/service/user/impl/UserServiceImpl.java]: why not moving it to user mapper?

File: src/main/java/bookstore/validation/FieldMatchValidator.java [hunk 19]
```diff
@@ -0,0 +1,14 @@
...
+import bookstore.dto.user.UserRegistrationRequestDto;
+import jakarta.validation.ConstraintValidator;
+import jakarta.validation.ConstraintValidatorContext;
+
+public class FieldMatchValidator implements ConstraintValidator<FieldMatch,
+        UserRegistrationRequestDto> {
+    @Override
+    public boolean isValid(UserRegistrationRequestDto requestDto,
+                           ConstraintValidatorContext constraintValidatorContext) {
+        return requestDto.password().equals(requestDto.repeatPassword());
```
Comment (NazarSmith228) [inline src/main/java/bookstore/validation/FieldMatchValidator.java]: what if either of these fields is null?

**Review by NazarSmith228 — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution

File: src/main/java/bookstore/dto/user/UserRegistrationRequestDto.java (hunk 4 above)
Comment (Cayman4010) [reply src/main/java/bookstore/dto/user/UserRegistrationRequestDto.java:23]: I think it`s not a required field

File: src/main/java/bookstore/exception/CustomGlobalExceptionHandler.java (hunk 5 above)
Comment (Cayman4010) [reply src/main/java/bookstore/exception/CustomGlobalExceptionHandler.java]: It`s magic!) I thought it was a very big row

File: src/main/java/bookstore/service/user/UserService.java (hunk 16 above)
Comment (Cayman4010) [reply src/main/java/bookstore/service/user/UserService.java:8]: Then I won't be able to throw an error in UserServiceImpl when registering with an existing email

File: src/main/java/bookstore/service/user/impl/UserServiceImpl.java (hunk 17 above)
Comment (Cayman4010) [reply src/main/java/bookstore/service/user/impl/UserServiceImpl.java:27]: Then I won't be able to throw an error when registering with an existing email

File: pom.xml [hunk 20]
```diff
@@ -72,6 +73,29 @@
             <artifactId>springdoc-openapi-starter-webmvc-ui</artifactId>
             <version>2.2.0</version>
         </dependency>
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-security</artifactId>
+            <version>3.1.5</version>
```
Comment (NazarSmith228) [inline pom.xml]: you may skip the version - it is taken from `spring-boot-starter-parent`

File: src/main/java/bookstore/config/SecurityConfig.java [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+                .addFilterBefore(jwtAuthenticationFilter,
+                        UsernamePasswordAuthenticationFilter.class)
+                .userDetailsService(userDetailsService)
+                .build();
+    }
+
+    @Bean
+    public AuthenticationManager authenticationManager(
+            AuthenticationConfiguration authenticationConfiguration
+    ) throws Exception {
```
Comment (NazarSmith228) [inline src/main/java/bookstore/config/SecurityConfig.java]:
```suggestion
            AuthenticationConfiguration authenticationConfiguration) throws Exception {
```

File: src/main/java/bookstore/dto/user/UserLoginRequestDto.java [hunk 22]
```diff
@@ -0,0 +1,15 @@
...
+import jakarta.validation.constraints.Size;
+
+public record UserLoginRequestDto(
+        @NotBlank
+        @Email
+        String email,
+        @NotBlank
+        @Size(min = 4, max = 60)
+        String password
+) {
```
Comment (NazarSmith228) [inline src/main/java/bookstore/dto/user/UserLoginRequestDto.java]:
```suggestion
        String password) {
```

File: src/main/java/bookstore/dto/user/UserResponseDto.java [hunk 23]
```diff
@@ -0,0 +1,10 @@
+package bookstore.dto.user;
+
+public record UserResponseDto(
+        Long id,
+        String email,
+        String firstName,
+        String lastName,
+        String shippingAddress
+) {
```
Comment (NazarSmith228) [inline src/main/java/bookstore/dto/user/UserResponseDto.java]:
```suggestion
        String shippingAddress) {
```

File: src/main/java/bookstore/model/RoleName.java [hunk 24]
```diff
@@ -0,0 +1,6 @@
+package bookstore.model;
+
+public enum RoleName {
```
Comment (NazarSmith228) [inline src/main/java/bookstore/model/RoleName.java]: this can be nested in `Role` class

File: src/main/java/bookstore/model/User.java (hunk 9 above)
Comment (NazarSmith228) [reply src/main/java/bookstore/model/User.java]: @Cayman4010 you still have `EAGER` here :)

File: src/main/java/bookstore/model/RoleName.java [hunk 25]
```diff
@@ -0,0 +1,6 @@
...
+    USER,
+    ADMIN
```
Comment (NazarSmith228) [inline src/main/java/bookstore/model/RoleName.java]: as far as I remember `hasRole` automatically adds `ROLE_` prefix to the arguments

it means that `hasRole("USER")` with check for `ROLE_USER` role - but here you have just `USER`

pls double check whether everything works with current setup and let me know in the chat :)

**Review by NazarSmith228 — CHANGES_REQUESTED** [review]
Good job! Just a couple of remarks left

File: src/main/java/bookstore/model/RoleName.java (hunk 25 above)
Comment (Cayman4010) [reply src/main/java/bookstore/model/RoleName.java]: I did some refactoring and now it works

File: pom.xml [hunk 26]
```diff
@@ -130,6 +153,11 @@
             <plugin>
                 <groupId>org.liquibase</groupId>
                 <artifactId>liquibase-maven-plugin</artifactId>
+                <configuration>
+                    <url>jdbc:mysql://localhost:3306/bookstore</url>
+                    <username>root</username>
+                    <password>75877587</password>
+                </configuration>
```
Comment (liliia-ponomarenko) [inline pom.xml]: Let’s remove it from pom.xml

**Review by liliia-ponomarenko — APPROVED** [review]
Great! But fix my comment before merging ;)

**Pull request merged**